### ISSUE TITLE
[CODEX] attempting to delete the last instruction of a program causes CODEX to lock up

### ISSUE THREAD

**Issue by jestin** [X16Community/x16-rom#31]
This can be easily reproduced by creating a new program, and entering in a few instructions above the default `RTS`.  Then you simply delete that default RTS (it does not matter if the label is deleted or modified) and the whole screen will freeze.

![image](https://user-images.githubusercontent.com/669174/221268773-f167392a-300e-4bac-8776-0a40d1a4a66a.png)

The above screen shot is how it appears after hitting F3 for `-INS` on the last line.  After this, there was nothing to do but reboot with  RUN/STOP+RESTORE (ctrl-R on the emulator).

**Comment by stefan-b-jakobsson** [comment]
I don't know much about CODEX; can barely use it at all.

But the problem you describe seems to be in codex/src/edit.s, function edit_delete.

A bit down in that function, the kernel routine MEMCOPY (=memory_copy) is called. When you're deleting the last line, it is called with an unfortunate param - the size of the copy is 0xffff, probably messing up all of the X16 memory.

And this seems to be caused by an attempt to do some code optimization at line 132, where it's said in a comment that the size is end - start + 1. This is a 16 bit subtraction of R1-R0 with carry clear. I suppose the author thought that by clearing the carry you would get +1, but in reality you will get a further -1.

I tested to replace the line 132 CLC with a SEC. And I then put a 16 bit addition after line 137:
```
        clc
	lda	 r2L
	adc	 #1
	sta	 r2L
	lda	 r2H
	adc	 #0
	sta	 r2H
```

After this I could delete both the last line and a preceding line without crashes.

**Comment by stefan-b-jakobsson** [comment]
I think there might be more issues with CODEX. It got it locked up in a different situation, but did not record exactly what I was doing.

**Comment by jestin** [comment]
There are definitely more problems.  This issue is just the one that I remembered what I was doing and was able to reproduce.

I actually contacted the developer earlier this week to ask a question, and he said he'd try to look into things this weekend.  I told him about the new repo, but I'm not sure what his plan is.

**Comment by mooinglemur** [comment]
Closed by #32
